### PR TITLE
Add AstNode::foreach method for simple pre-order traversal

### PR DESCRIPTION
**Pull request by gezalore**
As discussed, please see how you like it.

### CODE REVIEW

File: src/V3Ast.h [hunk 1]
```diff
@@ -1921,11 +1931,110 @@ class AstNode VL_NOT_FINAL {
...
 #include "V3Ast__gen_impl.h"  // From ./astgen
 
+// Specializations of privateMayBeUnder
+template <> inline bool AstNode::mayBeUnder<AstCell>(const AstNode* nodep) {
+    return !VN_IS(nodep, NodeStmt) && !VN_IS(nodep, NodeMath);
+}
+template <> inline bool AstNode::mayBeUnder<AstNodeAssign>(const AstNode* nodep) {
+    return !VN_IS(nodep, NodeMath);
+}
+
```
Comment (wsnyder) [inline src/V3Ast.h:2038]: Ideally I think we want this in the AstCell etc class definition, then ast gen to process something to make these, so it's closer to where people want to read it.  I'm ok waiting to do that.

File: src/V3MergeCond.cpp [hunk 2]
```diff
@@ -337,7 +319,7 @@ class MergeCondVisitor final : public VNVisitor {
         m_mgCondp = nullptr;
         m_mgLastp = nullptr;
         m_mgNextp = nullptr;
-        m_markVars.clear();
+        AstNode::user1ClearTree();  // Clear marked variables
         AstNode::user2ClearTree();
```
Comment (wsnyder) [inline src/V3MergeCond.cpp:323]: Unrelated, for a future cleanup I think AstNode::user1ClearTree should instead be a clear operation executed on the user1inuse object. Then it's clear that the code has "permission" to do the clear.   (The user1ClearTree function far predates tho

File: src/V3MergeCond.cpp [hunk 3]
```diff
@@ -407,7 +389,8 @@ class MergeCondVisitor final : public VNVisitor {
             m_mgFirstp = nodep;
             m_mgCondp = condp;
             m_listLenght = 0;
-            m_markVars.mark(condp);
+            // Mark variable references in the condition
+            condp->foreach<AstVarRef>([](const AstVarRef* nodep) { nodep->varp()->user1(1); });
```
Comment (wsnyder) [inline src/V3MergeCond.cpp:393]: This is so much better to read!!

File: src/V3Premit.cpp [hunk 4]
```diff
@@ -37,52 +37,6 @@
...
-    // METHODS
-    VL_DEBUG_FUNC;  // Declare debug()
-
-    // VISITORS
-    virtual void visit(AstNodeAssign* nodep) override {
-        // AstNode::user3ClearTree();  // Implied by VNUser3InUse
-        // LHS first as fewer varrefs
-        iterateAndNextNull(nodep->lhsp());
-        // Now find vars marked as lhs
-        iterateAndNextNull(nodep->rhsp());
```
Comment (wsnyder) [inline src/V3Premit.cpp:61]: I think you changed behavior here.  The old code did LHS then RHS, the new code will do RHS than LHS.

File: src/V3Task.cpp [hunk 5]
```diff
@@ -581,13 +525,7 @@ class TaskVisitor final : public VNVisitor {
             refp->taskp()->fvarp()->user2p(outvscp);
         }
         // Replace variable refs
-        // Iteration requires a back, so put under temporary node
-        {
-            AstBegin* const tempp = new AstBegin(beginp->fileline(), "[EditWrapper]", beginp);
```
Comment (wsnyder) [inline src/V3Task.cpp:586]: How did you get rid of having the AstBegin wrapper?

Similar below.

**Review by wsnyder — COMMENTED** [review]
This is a great cleanup.  Only questions where it looks like it wasn't a direct replacement - might be worth you also reviewing manually the old versus new again too (if you didn't already).

File: src/V3Ast.h (hunk 1 above)
Comment (gezalore) [reply src/V3Ast.h:2038]: Agreed, I would go one further and strengthen the accessors/constructors so you can prove these relationships statically, which astgen can then extract.

File: src/V3MergeCond.cpp (hunk 3 above)
Comment (gezalore) [reply src/V3MergeCond.cpp:393]: BTW all these are inspired by the Scala standard library collections, which also have `exists`, `forall`, `collect`, etc. Sadly does not fit in C++ just as nicely.

File: src/V3Task.cpp (hunk 5 above)
Comment (gezalore) [reply src/V3Task.cpp:586]: Only `AstVarRef` nodes are replaced, and none of the input nodes are that (i.e.: all `AstVarRef` are under the input list, so this works just fine. `backp()` is not required for `foreach`

File: src/V3Premit.cpp (hunk 4 above)
Comment (gezalore) [reply src/V3Premit.cpp:61]: If you look at the replacement at the use site, this visitor is always invoked on `AstNodeAssign`, and I replaced it with 2 separate `foreach`, first on the LHS, then or the RHS. Strictly, the behaviour is changed if there is a write ref on the RHS that is subsequently also read, but that was kind of bust in the original too, so I think this is ok.

**Comment by gezalore** [comment]
> might be worth you also reviewing manually the old versus new again too (if you didn't already).

I did, this was not an automatic thing. There are even a few left that could be made to work but I'm not sure they are worth it.

File: src/V3Ast.h (hunk 1 above)
Comment (wsnyder) [reply src/V3Ast.h:2038]: BTW proving statically will nicely make sure "new" works, but most of the hard bugs are due to bad edits, which don't call the accessors/constructors (e.g. lots of replaceWith's), so we must always have V3Broken do a dynamic check regardless.

**Comment by gezalore** [comment]
This does break some ext tests, will look into it shortly

**Pull request merged**